Automattic/newspack-popups
# proof of concept: segmentation criteria system

### All Submissions:






### Changes proposed in this Pull Request:

pemhSX-jU-p2

This is a proof of concept of how the segmentation criteria can be abstracted into a system that allows criteria registration and the use of custom or common matching functions. The common matching function will work according to its criteria type and matching attribute in the reader data library store.

The script is currently being enqueued and creates a sample segmentation config with the following criteria config values for matching:

```json
{
  "articles_read": {
    "min": 1,
    "max": 10
  },
  "articles_read_in_session": {
    "min": 1,
    "max": 10
  },
  "newsletter": {
    "value": 1
  },
  "referrer_sources": {
    "value": "google.com,facebook.com"
  }
}
```

It can be tested by populating the reader data library store with values expected by the config.

### How to test the changes in this Pull Request:

1. Make sure you have newspack-plugin updated with the most recent master (to include https://github.com/Automattic/newspack-plugin/pull/2451)
2. Open the site in incognito and open the browser console
3. Confirm you see the following logged:
     ```
     {criteriaId: 'articles_read', matched: false}
     {criteriaId: 'articles_read_in_session', matched: false}
     {criteriaId: 'newsletter', matched: false}
     {criteriaId: 'referrer_sources', matched: false}
     ```
4. Push a manual `article_view` activity to satisfy the `articles_read` criteria by entering the following in the console:
     ```js
     newspackRAS.push( [ 'article_view' ] );
     ```
5. Refresh and confirm the `articles_read` and  `articles_read_in_session` criteria has `matched` as `true`
6. Do the same for the other criteria:
     ```js
     newspackReaderActivation.store.set( 'is_subscriber', true );
     newspackReaderActivation.store.set( 'referrer', 'google.com' );
     ```
7. Refresh and confirm all logged entries have `matched` as `true`

### Other information:
